thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+
+\medskip
+All of the dependencies are bound to a specific minor version. Doing so ensures that in case one of the vendors implements a breaking change our application and pipelines do not break.
+
+\subsubsection{Frontend}
+
+The frontend code is linted on every commit using \tech{Husky}, \tech{ESLint}, and \tech{Prettier}. It is not possible to push an incorrectly formatted code to the repository.
+Uniformly formatted code improves the clarity, allowing a code reviewer to focus on business logic and potential bugs instead of code-formatting.
+
+\medskip
+Most of the business logic on the frontend is tested. The test runner of choice is \tech{Jest}.
+
+\medskip
+
+To aid the code review process, every pull request with changes related to the frontend component is deployed as a \tech{preview application} on \tech{Netlify}. A \tech{preview application} is an isolated environment with an application built using the code submitted along with a pull request.
+
+\medskip
+
+Running the whole test suite is as simple as executing \texttt{yarn test}.
+
+\begin{figure}[H]
+  \caption{Workflow - frontend component. Every commit is tested and built.}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-frontend-pipeline.png}
+\end{figure}
+
+\subsubsection{Solver}
+
+The Solver is a library written in python and tested using the \tech{unittest} module. Most of the business logic is covered. 
+
+The line test coverage is 77\%.
+
+\begin{figure}[H]
+  \caption{Workflow - solver component. Every change is tested.}
+  \centering
+    \includegraphics[width=0.8\textwidth]{assets/3-solver-pipeline.png}
+\end{figure}
+
+
+\subsubsection{Community}
+
+The community component is written in \tech{Python} using the \tech{Django} framework and therefore uses the \tech{unittest} module that is built-in to the Python standard library. All of the public-facing endpoints are fully covered with integration tests verifying that authentication, authorization, and request body handling are properly implemented. The line test coverage is 96\%.
+
+\medskip
+Code style abides by the \tech{PEP8} standard.
+
+\medskip
+To ensure build reproducibility, the community component is enclosed in a Docker image, which is automatically deployed to a production environment. 
+
+\begin{figure}[H]
+  \caption{Pipeline - Community backend component.}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-community-pipeline.png}
+\end{figure}
+
+\subsection{Deployment}
+
+Origuide uses Docker images for packaging components. Images are pushed to the \tech{Docker Hub} repository where they are publicly available.
+
+\medskip
+The deployment configuration is specified using \tech{docker-compose} files and is environment agnostic due to having environment configurations externalized to \texttt{.env} files. This makes \tech{Origuide}'s backend deployable using a single command.
```

Comment:
Also, the long first sentence sounds overcomplicated. I'd split it in 2.

Deployment configuration is specified using \tech{docker-compose} files. Because all the environment specific configuration is externalized to .env files, deployment configuration remains environment agnostic.